stan/math/gpu/kernels/matrix_multiply.hpp

```diff
@@ -0,0 +1,89 @@
+#ifndef STAN_MATH_GPU_KERNELS_MATRIX_MULTIPLY_HPP
+#define STAN_MATH_GPU_KERNELS_MATRIX_MULTIPLY_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/gpu/kernel_cl.hpp>
+
+namespace stan {
+namespace math {
+namespace opencl_kernels {
+// \cond
+const char* matrix_multiply_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Matrix multiplication on the GPU
+     *
+     * @param[in] A the left matrix in matrix multiplication
+     * @param[in] B the right matrix in matrix multiplication
+     * @param[out] C the output matrix
+     * @param[in] M Number of rows for matrix A
+     * @param[in] N Number of rows for matrix B
+     * @param[in] K Number of cols for matrix A and number of rows for matrix B
+     */
+    __kernel void matrix_multiply(
+        const __global read_only double* A, const __global read_only double* B,
+        __global write_only double* C, const read_only int M,
+        const read_only int N, const read_only int K) {
+      // workitem index inside the workgroup
+      const int workgroup_row = get_local_id(0);
+      const int workgroup_col = get_local_id(1);
+      // global workitem index
+      const int i = THREAD_BLOCK_SIZE * get_group_id(0) + workgroup_row;
+      const int j = THREAD_BLOCK_SIZE * get_group_id(1) + workgroup_col;
+
+      // local memory
+      __local double A_local[THREAD_BLOCK_SIZE][THREAD_BLOCK_SIZE];
+      __local double B_local[THREAD_BLOCK_SIZE][THREAD_BLOCK_SIZE];
+
+      double acc[WORK_PER_THREAD_MULT];
+      for (int w = 0; w < WORK_PER_THREAD_MULT; w++) {
+        acc[w] = 0.0;
+      }
+
+      const int num_tiles = (K + THREAD_BLOCK_SIZE - 1) / THREAD_BLOCK_SIZE;
+      // iterate over all tiles
+      for (int tile_ind = 0; tile_ind < num_tiles; tile_ind++) {
+        // each workitem copies WORK_PER_THREAD_MULT values to the local
+        // memory
+        for (int w = 0; w < WORK_PER_THREAD_MULT; w++) {
+          const int tiled_i = THREAD_BLOCK_SIZE * tile_ind + workgroup_row;
+          const int tiled_j = THREAD_BLOCK_SIZE * tile_ind + workgroup_col;
+
+          A_local[workgroup_col + w * THREAD_BLOCK_SIZE_MULT_COL][workgroup_row]
+              = A[(tiled_j + w * THREAD_BLOCK_SIZE_MULT_COL) * M + i];
```

Comment:
I think we could use the macro for the matrices here if we change them to something like
```cpp
#define A(i, j, k) A[j * k + i]
```

Though I think I'd like to wait till another PR to do that